### PR TITLE
[do not merge] proposal for new backup storage location layout

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <steve@heptio.com>

We've discussed this across a few different issues now, and I wanted to put a formal proposal out for us all to review.  I think I'd like to do this for v0.10, as it will address #576 and will also be used by #825.  Please provide your input on both the content of the proposal and the timing. Thanks!

### CODE REVIEW

File: docs/PROPOSAL-backup-location-layout.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      keys/
+      locks/
+      snapshots/
+    ...
+```
+
+# Revised Layout
+
+The revised layout adds top-level "directories" within the root of the backup storage location to organize the contents.
+The initial set of top-level directories is:
```
Comment (nrb) [inline docs/PROPOSAL-backup-location-layout.md]: I would add how the concept of ownership and read/write access factors into each directory here.

e.g. I expect only the owner to write to `backups` and `restic`. All other clusters could read all data, and potentially write to `restores` and possibly `metadata`, such as when claiming ownership after a migration.

File: docs/PROPOSAL-backup-location-layout.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+There are two main approaches I see for migrating to the new layout:
+- **Manual Migration** - the Ark codebase gets modified to support only the new layout. As part of the release containing this change,
+we provide a set of script templates, one per provider, to rearrange existing backup storage locations to conform to the new layout.
+These scripts would execute a series of commands (e.g. `aws s3 cp` for S3) to move existing directories/files around.
+- **Backwards Compatibility** - the Ark codebase is modified such that new backups/restores/etc. are stored within the new layout structure,
+but existing backup data stored directly in the root of the backup storage location is still recognized and supported as the basis for restores,
+delete operations, etc. Over time, all backups directly in the root should eventually be garbage-collected and only the new layout will remain.
+Additionally, we would probably remove all traces of the backwards-compatibility code as of Ark v1.0.
+
+My preferred approach is **manual migration**. We're still pre-1.0 so we don't make any guarantees about backwards compatibility. Any code that
```
Comment (nrb) [inline docs/PROPOSAL-backup-location-layout.md]: I think I'd prefer manual migration, too, with the caveat that I think we will see quite a few questions from people shortly after whichever release contains this change happens. I think people trying to restore from backups made prior to this layout will be most affected, especially if they do a cluster migration and happen to be using the `latest` image tag. But, as is stated, we're currently not guaranteeing backwards compat.

We should probably have the instructions for migrating in the docs *and* front and center on our announcements.

File: docs/PROPOSAL-backup-location-layout.md (hunk 1 above)
Comment (skriss) [reply docs/PROPOSAL-backup-location-layout.md]: Will do. I think it probably makes the most sense for restore logs/results of backups not owned by the target cluster to just go into the target cluster's default location under restores/ - WDYT?

Comment (skriss) [reply docs/PROPOSAL-backup-location-layout.md]: Added some info on this.  Gonna punt on deciding where restore logs go 'til next release.

File: docs/PROPOSAL-backup-location-layout.md (hunk 2 above)
Comment (skriss) [reply docs/PROPOSAL-backup-location-layout.md]: We could probably add some code that detects if a location is using the old layout and if so, log something pointing the user to a migration script and terminate.

Comment (nrb) [reply docs/PROPOSAL-backup-location-layout.md]: +1 to that.

Honestly I don't think we can overcommunicate when changes like this happen - there'll always be someone who doesn't read the messages or errors, but plenty of pointers to the process should help.

**Comment by skriss** [comment]
@carlisia @nrb @wwitzel3 @ncdc I'd like to decide soon if this layout makes sense, and if we want to do this as part of v0.10.  I'm in favor of both.  In particular, implementing this layout is going to make restic work related to locations much easier (i.e. sharing a bucket/prefix with the main Ark store, allowing multiple locations) and makes the sync controller improvements easier since we can put the `lastmodified` file into `metadata/`. 

Re: migration, I was thinking that the migration scripts can be really simple if we just move everything that's currently in the root of an Ark bucket into `backups/` in the new layout.  This means that any existing restore files won't go into `restores/`, so won't be accessible via CLI, but are still around.  I think this is fine since these are kind of ephemeral anyway and really only provide value immediately after a restore.  We could do the restore file migration too but the scripts will just get more complicated.  WDYT?

**Comment by ncdc** [comment]
@skriss I'm probably going to be 100% heads down on other things this week. If I have time to come up for air, I'll review & respond. Otherwise, it'll be next week from me.

**Comment by rosskukulinski** [comment]
I'm on board with manual migration, but we should definitely be clear in release notes, email to group, and I like the idea of exiting Ark Server if the storage layout doesn't match what it expects.

**Comment by skriss** [comment]
I'm closing this PR - any future discussion can move to #853

**Pull request closed without merging**